lib/http/serve/dir.go

```diff
@@ -146,82 +148,86 @@ func (d *Directory) ProcessQueryParams(sortParm string, orderParm string) *Direc
 	d.Sort = sortParm
 	d.Order = orderParm
 
-	var toSort sort.Interface
-
+	var sortFn func(a, b DirEntry) int
 	switch d.Sort {
 	case sortByName:
-		toSort = byName(*d)
+		sortFn = sortDirEntryByName
 	case sortByNameDirFirst:
-		toSort = byNameDirFirst(*d)
+		sortFn = sortDirEntryByNameDirFirst
 	case sortBySize:
-		toSort = bySize(*d)
+		sortFn = sortDirEntryBySize
 	case sortByTime:
-		toSort = byTime(*d)
+		sortFn = sortDirEntryByTime
 	default:
-		toSort = byNameDirFirst(*d)
-	}
-	if d.Order == "desc" && toSort != nil {
-		toSort = sort.Reverse(toSort)
+		sortFn = sortDirEntryByNameDirFirst
 	}
-	if toSort != nil {
-		sort.Sort(toSort)
+	slices.SortFunc(d.Entries, sortFn)
+
+	if d.Order == "desc" {
+		slices.Reverse(d.Entries)
 	}
 
 	return d
-
 }
 
-type byName Directory
-type byNameDirFirst Directory
-type bySize Directory
-type byTime Directory
+func sortDirEntryByName(a, b DirEntry) int {
+	aLeaf := strings.ToLower(a.Leaf)
+	bLeaf := strings.ToLower(b.Leaf)
 
-func (d byName) Len() int      { return len(d.Entries) }
-func (d byName) Swap(i, j int) { d.Entries[i], d.Entries[j] = d.Entries[j], d.Entries[i] }
+	// trim trailing '/' because it's not part of a filename. This also ensures
+	// correct dir order - 'test/' should be placed before 'test 1/'
+	if a.IsDir {
+		aLeaf = strings.TrimSuffix(aLeaf, "/")
+	}
+	if b.IsDir {
+		bLeaf = strings.TrimSuffix(bLeaf, "/")
+	}
 
-func (d byName) Less(i, j int) bool {
-	return strings.ToLower(d.Entries[i].Leaf) < strings.ToLower(d.Entries[j].Leaf)
+	return cmp.Compare(aLeaf, bLeaf)
```

Comment:
The sorting can be improved (?) even further by using `golang.org/x/text/collate`:

```diff
+var undCollator = collate.New(language.Und, collate.IgnoreCase, collate.Numeric)

func sortDirEntryByName(a, b DirEntry) int {
-	aLeaf := strings.ToLower(a.Leaf)
-	bLeaf := strings.ToLower(b.Leaf)
+	aLeaf := a.Leaf
+	bLeaf := b.Leaf

	...

-	return cmp.Compare(aLeaf, bLeaf)
+	return undCollator.CompareString(aLeaf, bLeaf)
}
```

This code will sort numbers within filenames in natural order: `test 1`, `test 2`, `test 12` instead of `test 1`, `test 12`, `test 2`. There are some drawbacks:

- binary size will be increased by ~1.36 MiB
- `collate.Collator.CompareString` is significantly slower